ystia/yorc
# Bugfix/gh 286 bootstrap new version

# Pull Request description

Fixed an issue where after an update of yorc binary from 3.1 to 3.2 in an environment used to bootstrap a setup, a new bootstrap was failing because it was re-using an alien4cloud distribution download stored under the working directory during the 3.1 bootstrap,
while the 3.2 bootstrap is requiring a different version of Alien4Cloud.

## Description of the change

### What I did

A bootstrap stores now versions of components it has downloaded in a new file `<work directory>/versions.yaml`.
Before attempting a download, the bootstrap command will verify if the current yorc binary component versions requirements (versions are hard-coded in the binary) are the same as the previous yorc binary that was used to download these components under the work directory (versions stored in `<work directory>/versions.yaml` if this file exists).
If requirements differ or previous requirements are unknown, the current bootstrap will download and overwrite bundles stored under the work directory.

### How to verify it

Download Yorc 3.1.2:
```
$ wget https://dl.bintray.com/ystia/yorc-engine/3.1.2/yorc-3.1.2.tgz
$ tar xzf yorc-3.1.2.tgz
```

Cleanup current environment:
```
$ ./yorc bootstrap cleanup
$ rm -Rf work
```
Bootstrap a setup :
```
$ ./yorc bootstrap -v <path to input yaml file>
```

Once the bootstrap has sucessfully finished, uninstall the setup and cleanup environment:
```
$ ./yorc d undeploy -p <deployment id>
$ ./yorc d info -f <deployment id>
$ ./yorc bootstrap cleanup
$ rm yorc
```

Download a yorc distribution containing changes in this pull request:
```
$ wget https://ystia.jfrog.io/ystia/binaries/ystia/yorc/dist/bugfix/GH-286-bootstrap-new-version/yorc-3.2.0-SNAPSHOT.tgz
$ tar xzf yorc-3.2.0-SNAPSHOT.tgz
```
Bootstrap a setup :
```
$ ./yorc bootstrap -v  <path to input yaml file> -n test
```
Check the bootstrap will finish successfuly. This bootstrapped setup is using Alien4Cloud 2.1.1 while the previous bootstrapped setup was using Alien4Cloud 2.1.0.


### Description for the changelog

Yorc bootstrap does not correctly treat default alien4cloud version download ([GH-286](https://github.com/ystia/yorc/issues/286))

## Applicable Issues

https://github.com/ystia/yorc/issues/286
